### PR TITLE
Allow for user feedback

### PR DESCRIPTION
**Pull request by ajbozarth**
This adds a UI for users to provide feedback on the Qiskit Code Assistant as well as specific prompts.

### Included features

- Feedback status bar icon: When clicked opens a feedback dialog for general feedback on the extension or model. The text field is immediately focused and and supports multiline input.
Screenshot 2024-10-30 at 5 14 43 PM 
Screenshot 2024-10-31 at 8 35 53 PM
- Prompt specific feedback: Once a prompt response has been received a feedback button is added to the Notebook cell toolbar. When clicked a feedback dialog connected to the most recently received prompt response is opened. The prompt response does not necessarily need to have been in the current cell, or even file. 
Screenshot 2024-10-30 at 5 14 54 PM
Screenshot 2024-10-30 at 5 15 21 PM
Screenshot 2024-10-31 at 8 35 46 PM
- After feedback is successfully sent to the service a toast notification is triggered 
Screenshot 2024-10-30 at 5 18 38 PM

### Current Limitations

1. The prompt feedback button is only available in notebooks and not the text editor.
    - This is due to not having a good location to put the button
    - **Workaround**: Due to the global context noted in limitation 2, users can switch to a notebook after receiving a prompt in the text editor and submit review via the feedback button.
2. The prompt feedback button is not tied to the current cell specifically and is just tied to the most recent prompt response returned to the current instance of jupyterlab.
    - This is a solvable limitation, but would require an excessive amount of code to address (would at least more than triple the amount of code added in this PR)
    - **Note**: This scope is still limited to the users current browser tab and not across all users connected to the current Jupyter lab instance
3. Thumbsup/thumbsdown buttons are not surfaced in the cell toolbar but will be in the dialog form.
    -  Adding these would be straight forward, but with the above limitation related to scope and context I believe it would be very confusing to the user what they're responding to.
4. If a user select a thumbsup/thumbsdown they can't de-select it
    - This is due to using a radio input for the form field, the field also still show the radio buttons to see which is selected since browsers convert unicode to emojis, which don't have a good "select" visual

### Open Questions

1. Is the issue of scope and context in limitation 2 a deal breaker or can well move forward with the prompt feedback button have a "global" context?
2. Is my judgment on issue 3 valid or do we still want to implement it with global context?

### CODE REVIEW

**Comment by ajbozarth** [comment]
@cbjuan @vabarbosa This is just an initial draft, but outside of fixing the dialog UI it's fully featured and complete. If you could take a look, I included a detailed description and called out specific questions I'd like answers to.

**Comment by ajbozarth** [comment]
> 2. The prompt feedback button is not tied to the current cell specifically and is just tied to the most recent prompt response returned to the current instance of jupyterlab.

After a worrying thought entered my head I tested how far the scope go on this and confirmed it is limited to a specific user instance of jupyterlab and not a server instance. Ie the scope is shared across every open notebook, but not between the same jupyterlab instance opened in another browser tab. Relatedly it does not persist on page reload.

**Comment by ajbozarth** [comment]
I have just pushed the final updates and updated the top description with new screen shots, descriptions, and limitations

**Comment by ajbozarth** [comment]
@cbjuan I just pushed the update to include the input/output values as now allowed by the service https://qiskit-code-assistant.quantum.ibm.com/docs#/Feedback/submit_feedback_feedback_post

**Review by cbjuan — APPROVED** [review]
Code looks good to me, so approving

I'd love to test it before publishing a new version with this. Also, could you take a look to this PR, @vabarbosa ?

**Comment by ajbozarth** [comment]
> I'd love to test it before publishing a new version with this.

@cbjuan so do you want to test this PR by building it locally or do you mean we should merge and release an rc for testing via pip install wheel file?

**Comment by cbjuan** [comment]
I was thinking on testing it locally, but you're right, probably is cleaner to publish it as a rc and test it from there. Let's do it after @vabarbosa review

File: src/service/feedback.tsx [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+import { ReactWidget } from '@jupyterlab/ui-components';
+
+import { postFeedback } from './api';
+import { getCurrentModel } from './modelHandler';
+import { lastPrompt } from '../QiskitCompletionProvider';
+import { feedbackIcon } from '../utils/icons';
+import { IFeedbackForm, IFeedbackResponse } from '../utils/schema';
+
+const toBool = (bool?: string): boolean | undefined => {
+  switch (bool) {
```
Comment (vabarbosa) [inline src/service/feedback.tsx:29]: not a big deal now but would it be better/future proof if the `bool` string is set to lowercase

File: style/base.css [hunk 2]
```diff
@@ -14,13 +14,40 @@
...
+  max-height: 200px;
+  border: solid thin;
+}
+
+.jp-qiskit-code-assistant-feedback-prompt textarea,
+.jp-qiskit-code-assistant-feedback-general textarea {
+  width: 30em;
+  height: 5em;
+  font-family: inherit;
+  font-size: inherit;
```
Comment (vabarbosa) [inline style/base.css:36]: should the textarea also define a background color (so can properly be updated when switching between light/dark themes)?

image

File: src/service/feedback.tsx [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+
+function getFeedbackBodyWidget(
+  prompt?: string
+): Dialog.IBodyWidget<IFeedbackForm> {
+  const bodyWidget = ReactWidget.create(
+    <>
+      {!!prompt && (
+        <>
+          <b>Completion</b>
+          <pre>{prompt.trim()}</pre>
```
Comment (vabarbosa) [inline src/service/feedback.tsx:114]: would this better off as a `textarea` with `readonly` attribute?

**Review by vabarbosa — COMMENTED** [review]
just a couple minor comments, but nothing blocking

File: src/service/feedback.tsx (hunk 1 above)
Comment (ajbozarth) [reply src/service/feedback.tsx:29]: perhaps this helper function's name is deceptive, this actually converts the value of a radio button defined on line 122/127 to a boolean. The values `"true"` and `"false"` are set on those lines and could actually be any string we want, I could have had them be `"thumbsup"` and `"thumbsdown"`, but I figured truthy values would be easier.

so no, the `"true"` and `"false"` here is a enum not a string representation of a boolean, and so any other values would be considered "wrong"

File: style/base.css (hunk 2 above)
Comment (ajbozarth) [reply style/base.css:36]: I'll take a look at what jupyterlab css class that `textarea` needs to inherit to support the theme. That's not something to address in our own css.

Comment (ajbozarth) [reply style/base.css:36]: I'm unsure how you got that, when I switch the theme it works as expected 
Screenshot 2024-11-07 at 12 50 29 PM

Comment (vabarbosa) [reply style/base.css:36]: this is what i see (i tried switching themes and switching back. also restarted jupyterlab)

https://github.com/user-attachments/assets/fd9ca2a8-c2c3-4922-8985-05518031d433

File: src/service/feedback.tsx (hunk 3 above)
Comment (ajbozarth) [reply src/service/feedback.tsx:114]: Just did a bit of research and no, a readonly texture is not considered a good was to show formatted text.

While I was at it I checked to see if there was a simple way I missed to insert a read-only CodeMirrorEditor here like I had originally planned to match the design doc, but I only found two solutions, to use a third party lib or write a bunch of code to allow the extension to use the jupyterlab editor factory to generate it (I gave this a try and it's turned out to be even more difficult that just adding a bunch of code)

File: style/base.css (hunk 2 above)
Comment (ajbozarth) [reply style/base.css:36]: Ah that last part might be it, this extension requires Jupyterlab 4.3 (though I'm unsure why that would cause this css bug)

Comment (ajbozarth) [reply style/base.css:36]: Just reverted my lab version to 4.1.6 as you're using and I see the same. So this is definitely due to using an unsupported version. So not a bug worth fixing

File: src/service/feedback.tsx (hunk 3 above)
Comment (vabarbosa) [reply src/service/feedback.tsx:114]: 👍🏽

File: style/base.css (hunk 2 above)
Comment (vabarbosa) [reply style/base.css:36]: ah ok.. good to know

**Review by vabarbosa — APPROVED** [review]
thanks!

File: style/base.css (hunk 2 above)
Comment (vabarbosa) [reply style/base.css:36]: updated to 4.3 and all looks good... thx

**Pull request merged**